plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,484 @@
+sendContractInboxMessage rsp i it = do
+    runCommand (sendContractEvent @t) ContractEventSource
+        $ ContractInboxEvent
+        $ ContractInboxMessage
+            { ciInstance = i
+            , ciIteration = it
+            , ciPayload = rsp
+            }
+
+-- | Get the current state of the contract instance
+lookupContractState ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member (Error SCBError) effs
+    )
+    => ContractInstanceId
+    -> Eff effs (ContractInstanceState t)
+lookupContractState instanceID = do
+    mp <- runGlobalQuery (Query.contractState @t)
+    case Map.lookup instanceID mp of
+        Nothing       -> throwError $ OtherError $ "Contract instance not found " <> tshow instanceID
+        Just (Last s) -> pure s
+
+-- | For a given contract instance, take the first message
+--   from the inbox and update the contract with it.
+processContractInbox ::
+    forall t effs.
+        ( Member (EventLogEffect (ChainEvent t)) effs
+        , Member (ContractEffect t) effs
+        , Member (Error SCBError) effs
+        , Member Log effs
+        )
+    => ContractInstanceId
+    -> Seq ContractResponse
+    -> Eff effs ()
+processContractInbox instanceID messages = do
+    logInfo $ "processContractInbox start for " <> render (pretty instanceID)
+    case Seq.viewl messages of
+        msg Seq.:< _ -> do
+            logInfo . render $ "The first message is: " <+> pretty msg
+            logInfo "Looking up the contract."
+            -- look up contract 't'
+            ContractInstanceState{csCurrentIteration, csCurrentState, csContractDefinition} <- lookupContractState @t instanceID
+            -- process the message
+            let payload = Contract.contractMessageToPayload msg
+            newState <- Contract.invokeContractUpdate_ csContractDefinition csCurrentState payload
+
+            -- send out the new requests
+            -- see note [Contract Iterations]
+            let newIteration = succ csCurrentIteration
+            sendContractStateMessages csContractDefinition instanceID newIteration newState
+            logInfo . render $ "Updated contract" <+> pretty instanceID <+> "to new iteration" <+> pretty newIteration
+        _ -> logInfo "No messages."
+    logInfo "processContractInbox end"
+
+-- | Check the inboxes of all contracts.
+processAllContractInboxes ::
+    forall t effs.
+        ( Member (EventLogEffect (ChainEvent t)) effs
+        , Member (ContractEffect t) effs
+        , Member (Error SCBError) effs
+        , Member Log effs
+        )
+    => Eff effs ()
+processAllContractInboxes = do
+    logInfo "processAllContractInboxes start"
+    ContractInboxState{cisInboxMessages} <- runGlobalQuery (Query.inboxMessages @t)
+    traverse_ (uncurry (processContractInbox @t)) (Map.toList cisInboxMessages)
+    logInfo "processAllContractInboxes end"
+
+-- TODO different return type to indicate that the contract
+-- does indeed exist
```

Comment:
Yeah, it seems like a bunch of the things that look stuff up by contract should return `Maybe` instead of throwing errors and let the caller decide whether or not it's okay.

Reply:
The caller can always use `handleError` to deal with errors (maybe we just need a more specific error type than `SCBError`)

Reply:
I guess the difference is that I would use `Maybe` for something where I really want the caller to make a choice about what to do about it, and `Error` for stuff that can bubble up. Intuitively I feel like "that contract was missing" is probably not something you want to bubble up, but I'm not sure.

Reply:
Also this is a vague heuristic feeling, maybe `handleError` with a nicer error type is fine.